--- source/common/common/win32/thread_impl.h
+++ source/common/common/win32/thread_impl.h
@@ -39,7 +39,7 @@
 public:
   // Thread::ThreadFactory
   ThreadPtr createThread(std::function<void()> thread_routine, OptionsOptConstRef options) override;
-  ThreadId currentThreadId() override;
+  ThreadId currentThreadId() const override;
 };
 
 } // namespace Thread
--- source/common/common/win32/thread_impl.cc
+++ source/common/common/win32/thread_impl.cc
@@ -22,9 +22,9 @@
         return 0;
       },
       this, 0, nullptr));
-  if (options && options.thread_priority_ &&
-      !SetThreadPriority(thread_handle_, *options.thread_priority_)) {
-    ENVOY_LOG_MISC(warn, "Could not set the thread priority to {}", *options.thread_priority_);
+  if (options && options->priority_ &&
+      !SetThreadPriority(thread_handle_, *options->priority_)) {
+    ENVOY_LOG_MISC(warn, "Could not set the thread priority to {}", *options->priority_);
   }
   RELEASE_ASSERT(thread_handle_ != 0, "");
 }
@@ -42,7 +42,7 @@
   return std::make_unique<ThreadImplWin32>(thread_routine, options);
 }
 
-ThreadId ThreadFactoryImplWin32::currentThreadId() {
+ThreadId ThreadFactoryImplWin32::currentThreadId() const {
   // TODO(mhoran): test this in windows please.
   return ThreadId(static_cast<int64_t>(::GetCurrentThreadId()));
 }
--- envoy/api/os_sys_calls.h
+++ envoy/api/os_sys_calls.h
@@ -1,17 +1,27 @@
 #pragma once
 
+#ifndef WIN32
 #include <sys/resource.h>
+#endif
 #include <sys/stat.h>
 
 #include <chrono>
 #include <memory>
 #include <string>
 #include <vector>
 
 #include "envoy/api/os_sys_calls_common.h"
 #include "envoy/common/platform.h"
 #include "envoy/common/pure.h"
 #include "envoy/network/address.h"
 
+#ifdef WIN32
+struct rlimit {
+  int rlim_cur;
+  int rlim_max;
+};
+#define RLIMIT_NOFILE 7
+#endif
+
 namespace Envoy {
 namespace Api {
--- source/extensions/common/aws/BUILD
+++ source/extensions/common/aws/BUILD
@@ -75,6 +75,7 @@
     ],
     deps = [
         "//envoy/common:pure_lib",
+        "//envoy/common:time_interface",
         "//source/common/common:cleanup_lib",
         "//source/common/common:lock_guard_lib",
         "//source/common/common:thread_lib",
